# Project: Exploring eBay Car Sales Data

In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The aim of this project is to clean the data and analyze the included used car listings.

In [174]:
import pandas as pd
import pprint

In [175]:
# Read the autos.csv CSV file into pandas
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [176]:
# Use the DataFrame.info() and DataFrame.head() methods to print 
# information about the dataframe
autos.head(3)

dateCrawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offerType   price   abtest vehicleType  yearOfRegistration    gearbox  \
0   Angebot  $5,000  control         bus                2004    manuell   
1   Angebot  $8,500  control   limousine                1997  automatik   
2   Angebot  $8,990     test   limousine                2009    manuell   

   powerPS   model   odometer  monthOfRegistration fuelType       brand  \
0      158  andere  150,000km                    3      lpg     peugeot   
1      286     7er  150,000km                    6   benzin         bmw   
2      102    golf   70,000km                    7   benzin  volkswagen   

  notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
0              nein  2016-03-26 00:00:00             0       79588   
1              nein  2016-04-04 00:00:00             0       71034   
2              nein  2016-03-26 00:00:00             0       35394   

              lastSeen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

In [177]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [178]:
# Use the DataFrame.columns attribute to print 
# an array of the existing column names.
columns = autos.columns
print(columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [179]:
# Change the name of columns form camelcase to snakecase
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  num_photos  postal_code  \
0     peugeot              nein  2016-03-26 00:00:00           0        79588   
1         bmw              nein  2016-04-04 00:00:00           0        71034   
2  volkswagen              nein  2016-03-26 00:00:00           0        35394   
3       smart              nein  2016-03-12 00:00:00           0        33729   
4        ford              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [180]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

We can see that the column price and odometer columns are numeric values stored as text. Therefore, we need to clean the data. 

There are a number of text columns where most of the values are the same, 
seller and offer_type

In [181]:
# Use DataFrame.rename() to rename the column "odometer" to "odometer_km"
autos = autos.rename({'odometer':'odometer_km'}, axis=1)

In [182]:
# Fix data in price column and collect it in int type.
autos['price'] = (autos['price'].str.replace('$','')
                                .str.replace(',','')
                                .astype(int))
# Check by printing a few of row.
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [183]:
# Fix data in odometer_km column and collect it in int type.
autos['odometer_km'] = (autos['odometer_km'].str.replace('km','')
                                .str.replace(',','')
                                .astype(int))
# Check by printing a few of row.
autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [184]:
autos = autos.drop(["seller", "offer_type"], axis=1)

In [185]:
autos['price'].unique().shape
autos['price'].describe()
autos['price'].value_counts().sort_index(ascending=False).head()

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price, dtype: int64

We will calculate the distribution of values in the date_crawled, ad_created, and last_seen columns (all string columns) as percentages

In [186]:
# distribution of values in the date_crawled column
(autos["date_crawled"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: date_crawled, dtype: float64

In [187]:
# distribution of values in the ad_created column
(autos["ad_created"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

In [188]:
# distribution of values in the last_seen column
(autos["last_seen"]
        .str[:10]
        .value_counts(normalize=True, dropna=False)
        .sort_index()
        )

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: last_seen, dtype: float64

In [189]:
# Use Series.describe() to understand the distribution of registration_year.
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

## Dealing with Incorrect Registration Year Data
One thing that stands out from the exploration we did in the last screen is that the registration_year column contains some odd values:

* The minimum value is 1000, before cars were invented

* The maximum value is 9999, many years into the future

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [190]:
registration_year_bool = autos['registration_year'].between(1900, 2016)
autos = autos[registration_year_bool]
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  num_photos  postal_code  \
0              nein  2016-03-26 00:00:00           0        79588   
1              nein  2016-04-04 00:00:00           0        71034   
2              nein  2016-03-26 00:00:00           0        35394   
3              nein  2016-03-12 00:00:00           0        33729   
4              nein  2016-04-01 00:00:00           0        39218   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

In [191]:
#  the distribution of the remaining values
autos['registration_year'].value_counts(normalize=True)

2000    0.069834
2005    0.062776
1999    0.062464
2004    0.056988
2003    0.056779
          ...   
1939    0.000021
1927    0.000021
1929    0.000021
1948    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

In [192]:
brands = autos['brand'].value_counts().sort_values(ascending=False)
brands.head(20)

volkswagen        10188
bmw                5284
opel               5195
mercedes_benz      4580
audi               4149
ford               3352
renault            2274
peugeot            1418
fiat               1242
seat                873
skoda               770
mazda               727
nissan              725
citroen             669
smart               668
toyota              599
sonstige_autos      526
hyundai             473
volvo               444
mini                415
Name: brand, dtype: int64

We want to select the top 20 brand in the dataset.

In [193]:
brand_dict = {}
for key in brands.index:
    brand = autos[autos['brand'] == key]
    brand_dict[key] = brand['price'].mean()

In [194]:
sorted_brand_dict = dict(sorted(brand_dict.items(), key=lambda x: x[1], reverse=True))
sorted_brand_dict

{'porsche': 44553.46757679181,
 'citroen': 44534.79671150971,
 'sonstige_autos': 39621.77946768061,
 'mercedes_benz': 30317.447816593885,
 'land_rover': 19108.091836734693,
 'jeep': 11434.75,
 'jaguar': 11176.197368421053,
 'mini': 10460.012048192772,
 'audi': 9093.65003615329,
 'bmw': 8334.645155185466,
 'ford': 7263.015811455847,
 'volkswagen': 6516.457597173145,
 'chevrolet': 6488.981751824817,
 'skoda': 6334.91948051948,
 'dacia': 5915.528455284553,
 'kia': 5789.351906158357,
 'hyundai': 5308.53911205074,
 'opel': 5252.61655437921,
 'toyota': 5115.33388981636,
 'volvo': 4757.108108108108,
 'nissan': 4664.891034482758,
 'seat': 4296.492554410081,
 'mazda': 4010.7716643741405,
 'suzuki': 3995.757042253521,
 'honda': 3988.0,
 'alfa_romeo': 3984.85534591195,
 'subaru': 3765.038095238095,
 'smart': 3542.706586826347,
 'mitsubishi': 3333.8005115089513,
 'lancia': 3246.3653846153848,
 'chrysler': 3229.443181818182,
 'saab': 3211.6493506493507,
 'peugeot': 3039.4682651622,
 'fiat': 2711.80

## pandas series constructor and pandas dataframe constructor


In [195]:
# pandas series constructor
car_series = pd.Series(sorted_brand_dict)
car_series

porsche           44553.467577
citroen           44534.796712
sonstige_autos    39621.779468
mercedes_benz     30317.447817
land_rover        19108.091837
jeep              11434.750000
jaguar            11176.197368
mini              10460.012048
audi               9093.650036
bmw                8334.645155
ford               7263.015811
volkswagen         6516.457597
chevrolet          6488.981752
skoda              6334.919481
dacia              5915.528455
kia                5789.351906
hyundai            5308.539112
opel               5252.616554
toyota             5115.333890
volvo              4757.108108
nissan             4664.891034
seat               4296.492554
mazda              4010.771664
suzuki             3995.757042
honda              3988.000000
alfa_romeo         3984.855346
subaru             3765.038095
smart              3542.706587
mitsubishi         3333.800512
lancia             3246.365385
chrysler           3229.443182
saab               3211.649351
peugeot 

In [196]:
# pandas dataframes constructor
car_df = pd.DataFrame(car_series, columns = ['mean_price'])
car_df

mean_price
porsche         44553.467577
citroen         44534.796712
sonstige_autos  39621.779468
mercedes_benz   30317.447817
land_rover      19108.091837
jeep            11434.750000
jaguar          11176.197368
mini            10460.012048
audi             9093.650036
bmw              8334.645155
ford             7263.015811
volkswagen       6516.457597
chevrolet        6488.981752
skoda            6334.919481
dacia            5915.528455
kia              5789.351906
hyundai          5308.539112
opel             5252.616554
toyota           5115.333890
volvo            4757.108108
nissan           4664.891034
seat             4296.492554
mazda            4010.771664
suzuki           3995.757042
honda            3988.000000
alfa_romeo       3984.855346
subaru           3765.038095
smart            3542.706587
mitsubishi       3333.800512
lancia           3246.365385
chrysler         3229.443182
saab             3211.649351
peugeot          3039.468265
fiat             2711.801127
lada             2502.896552
renault          2395.416447
daihatsu         1556.382114
trabant          1552.080000
rover            1528.338462
daewoo           1019.861111

In this project, we practiced applying a variety of pandas methods to explore and understand a data set on car listings.